### PR TITLE
Adding pricing plans and serialize them within designs

### PR DESCRIPTION
**Pull request by 2potatocakes**
This PR is not ready yet but I wanted to get some feedback and see what people thought about the new `pricing_plans` schema and how they'll be serialized _within_ the design schema as well as independently.

This PR adds `pricing_plans` as a new schema type within elasticsearch. Currently there are only 3 plans (2 for hosted and 1 for sites).

These pricing plans will also be serialized _within_ the `design` schema, I've also serialized the `categories` within the `design` schema as well.

I'm looking to see if anyone has any concerns with the approach or even the structure of the pricing plans themselves? Cheers in advance!

### CODE REVIEW

File: data/designs/hosted/avada-creative-445e4bf1-987b-4b72-a5f5-ca4bf31a0b7f.json [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    "url": "https://themeforest.net/item/avada-responsive-multipurpose-theme/2833226"
+  },
+  "author": {
+    "name": "ThemeFusion",
+    "url": "https://themeforest.net/user/themefusion"
+  },
+  "checkout_url": "https://build.envatohostedservices.com/sign_up?bundle_id=155&product_id=560&look=bundle-default",
+  "pricing_plans": [
+    "hosted_annually",
+    "hosted_monthly"
```
Comment (travo) [inline data/designs/hosted/avada-creative-445e4bf1-987b-4b72-a5f5-ca4bf31a0b7f.json]: Yep, yep, yep.

File: data/designs/sites/designs.json [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+      "name": null,
+      "url": null
+    },
+    "author": {
+      "name": "Envato Sites",
+      "url": "https://sites.envato.com/"
+    },
+    "checkout_url": "https://sites.envato.com/create-site/koala/default",
+    "pricing_plans": [
+      "sites_monthly"
```
Comment (travo) [inline data/designs/sites/designs.json]: Yep! And...yep!

File: data/shopfront/pricing_plans.json [hunk 3]
```diff
@@ -0,0 +1,50 @@
+[
+  {
+    "uuid": "26294886-122a-4006-aae6-66cc62b7de2f",
+    "slug": "hosted_monthly",
```
Comment (travo) [inline data/shopfront/pricing_plans.json]: Ooooh, I see what you did there. Yep!

File: lib/catalogue/designs/serializer.rb [hunk 4]
```diff
@@ -2,20 +2,40 @@ module Catalogue
...
+        end
+        categories.map { |category| Categories::Serializer.serialize(category) }
+      end
+
+      def serialized_pricing_plans(pricing_plan_slugs=[])
+        pricing_plans = PricingPlans::PricingPlan.all.select do |pricing_plan|
+          pricing_plan_slugs.include?(pricing_plan.slug)
+        end
+        pricing_plans.map { |pricing_plan| PricingPlans::Serializer.serialize(pricing_plan) }
+      end
```
Comment (travo) [inline lib/catalogue/designs/serializer.rb]: These methods are clearly pretty similar, and can literally hear the rush of folks clamouring to refactor. But steady-on I say, steady on!

File: lib/ingestion/schema.rb [hunk 5]
```diff
@@ -54,31 +59,24 @@ def design_mappings
...
-        },
-        'selling_points' => {
+        'pricing_plans' => {
           'type' => 'object',
-          'properties' => {
-            'monthly' => Types::TEXT,
-            'yearly'  => Types::TEXT,
-          }
-        },
+          'properties' => pricing_plan_mappings
```
Comment (travo) [inline lib/ingestion/schema.rb]: This cleans up the schema quite a bit.

**Comment by blimpage** [comment]
Do pricing plans need a `name` attribute? We're currently displaying a name for each plan on the homepage:

screen shot 2017-08-11 at 10 03 42 am

**Comment by 2potatocakes** [comment]
@blimpage - yeah I did wonder that too, I originally had it in there and then swapped it out for the slug, I was thinking perhaps the nomenclature could be determined on the front end from the `term` attribute? Happy to put it back in there though if it makes life easier for us?

**Comment by blimpage** [comment]
@2potatocakes Yeah, we're already running into trouble with using `term` as a `name` due to "year" vs. "annual". I reckon we'll need a `name`.

**Comment by 2potatocakes** [comment]
@blimpage - Do we call it `name: "Annually"` and `name: "Monthly"` then? feels weird

**Comment by travo** [comment]
I still like this.

**Comment by blimpage** [comment]
@2potatocakes naming should be "Monthly" and "Annual". Conflicting adjective/noun-to-adverb confusion 😛 . I've been meaning to push a branch to fix this on the shopfront.

File: lib/catalogue/designs/serializer.rb [hunk 6]
```diff
@@ -2,20 +2,44 @@ module Catalogue
...
+            raise "Unrecognised Category slug! '#{slug}'" if category.nil?
+          end
+        end
+        categories.map { |category| Categories::Serializer.serialize(category) }
+      end
+
+      def serialized_pricing_plans(pricing_plan_slugs=[])
+        pricing_plans = pricing_plan_slugs.map do |slug|
+          PricingPlans::PricingPlan.all.detect {|plan| plan.slug == slug}.tap do |plan|
+            raise "Unrecognised PricingPlan slug! '#{slug}'" if plan.nil?
```
Comment (blimpage) [inline lib/catalogue/designs/serializer.rb]: :+1: This validation is awesome!

File: lib/catalogue/designs/serializer.rb [hunk 7]
```diff
@@ -2,20 +2,44 @@ module Catalogue
...
+          pricing_plans:      serialized_pricing_plans(design.pricing_plans),
         }
       end
+
+      private
+
+      def serialized_categories(category_slugs=[])
+        category_slugs << "all" unless category_slugs.include?("all")
+        categories = category_slugs.map do |slug|
+          Categories::Category.all.detect {|category| category.slug == slug}.tap do |category|
```
Comment (blimpage) [inline lib/catalogue/designs/serializer.rb]: `Categories::Category.all` loads our category data from disk. If we're calling this for each of our ~800 designs, then we're loading that data from disk ~800 times every time we reindex. I know this doesn't run in a production environment so it doesn't need to be lightning fast, but we might look at caching this at some point if things start getting slow.

Comment (blimpage) [inline lib/catalogue/designs/serializer.rb]: Oooh and @dferdian just pointed out to me that this is inside the `category_slugs.map` loop, so we're actually loading category data from disk multiple times per design. 🐌 🐌 🐌 🐌

Comment (2potatocakes) [inline lib/catalogue/designs/serializer.rb]: totally agree 👍

Comment (blimpage) [inline lib/catalogue/designs/serializer.rb]: (not a blocker for this PR, just an observation of something we might wanna optimise in future)

Comment (2potatocakes) [inline lib/catalogue/designs/serializer.rb]: lemme take another look 👍  cheers guys

File: data/designs/hosted/avada-creative-445e4bf1-987b-4b72-a5f5-ca4bf31a0b7f.json [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+    "name": "Avada",
+    "url": "https://themeforest.net/item/avada-responsive-multipurpose-theme/2833226"
+  },
+  "author": {
+    "name": "ThemeFusion",
+    "url": "https://themeforest.net/user/themefusion"
+  },
+  "checkout_url": "https://build.envatohostedservices.com/sign_up?bundle_id=155&product_id=560&look=bundle-default",
+  "pricing_plans": [
+    "hosted_annual",
```
Comment (smlparry) [inline data/designs/hosted/avada-creative-445e4bf1-987b-4b72-a5f5-ca4bf31a0b7f.json]: Perhaps for another time, but I still think we should be more clear that these are defining relationships and are going to be interpolated. 
Some visual indicator would be nice `>hosted_annual` ?

**Comment by travo** [comment]
I _still_ like this. Power of work @2potatocakes — love it. Have a 🍺

File: lib/catalogue/designs/serializer.rb (hunk 7 above)
Comment (smlparry) [inline lib/catalogue/designs/serializer.rb]: Could just extract it to a variable on line 26

File: spec/catalogue/designs/serializer_spec.rb [hunk 9]
```diff
@@ -40,5 +35,48 @@
         expect(serialized_design[:uuid]).to eq(uuid.downcase)
       end
     end
+
+    context "nested objects" do
+      context "categories" do
+        let(:all_category) {
```
Comment (smlparry) [inline spec/catalogue/designs/serializer_spec.rb]: perhaps just a personal preference thing, but for readability sake, i quite like using `do ... end` for multiline `let`'s

Totally up to you though

**Comment by 2potatocakes** [comment]
@blimpage I just added a `name` attribute for the pricing plans as mentioned earlier and named the 2 plans "Monthly", "Annual" for hosted and "Monthly" for the sites plans. Hope that's ok with everyone?

**Pull request merged**